### DISCUSSION TITLE
Outline material and material overrides

### DISCUSSION THREAD

**Discussion by Orctonk** [google/filament#6015] [Q&A]
Hi! I am currently working on implementing an outline shader which highlights interior edges as well as exterior edges ([See this link for reference](https://alexanderameye.github.io/notes/edge-detection-outlines/)). But I am running into some problems when it comes to accessing the required data. 

Given that the rendered scene is already has setup renderables which generate the required data (normals and depth) I would ideally like to be able to access this data somehow but from what I can tell there is no way to "hook into" the PostProcessing system or similar.

My second approach would thus be to render the scene from scratch and use MRT to generate textures with normals + depth data. This would however require me to setup a separate scene with new renderables containing the new material. I'm thus wondering if there is a way to render an existing scene using a new material for all objects or some cheap way to copy the scene and set new materials.

Do you have any recommendations for or experience with implementing these types of effects? Thanks in advance!

**Comment by Orctonk** [comment]
After some further searching I found a related issue [#415](https://github.com/google/filament/issues/415) which mentioned that a material override feature existed but was removed. I have for now implemented the system with double renderables as mentioned by @pixelflinger but it feels kind of cumbersome and hacky to have to construct these additional renderables since the outline shading is meant to be toggleable in my use case and since renderables are created in different places.

Considering that the original issue is quite old, have there been any developments since to make this effect easier to achieve?

**Reply by romainguy** [reply]
Material override was meant to override the materials of *all* the renderables in the scene, it would not help you. A post processing implementation would be best and we'd like to eventually expose the frame graph APIs to give access to the required buffers. @pixelflinger
